### ISSUE TITLE
I'm needing one of my Claim JProperties to have a value that is a JArray; My Claims are being double serialized.

### ISSUE THREAD

**Issue by kalanbates** [IdentityServer/IdentityServer3#1997]
The TL;DR is: I need a custom claim to have a value that is a JArray. The parser is double serializing my Claim value and I don't know how best to instruct it to not parse a particular claim.  Can you point me to any examples?

---
I need one of my claims to have a value that is a JArray.  
```C#
public async override Task GetProfileDataAsync(ProfileDataRequestContext context)
{
   var claims = new List<Claim>();
   claims.Add(new Claim("test", @"[{""name"":""foo""},{""name"":""bar""}]"));
   context.IssuedClaims = claims.AsEnumerable<Claim>();
}
```
What I need this to spit out is:
`"test": [{"name":"foo"},{"name":"bar"}],`
but pre-formatting this value as valid JSON causes the  parser to double serialize. So the profile endpoint spits out
`"test": "[{\"name\":\"foo\"},{\"name\":\"bar\"}]",`

Do you have any samples you can point me towards? 
Thanks.

**Comment by leastprivilege** [comment]
We don't have good support for that. But probably something we can fix.

How do you expect that claim to de-serialize "on the other end" - are you using .NET ?

**Comment by leastprivilege** [comment]
actually not sure if this can be easily done using the MS JWT library. investigating.

**Comment by leastprivilege** [comment]
You can take control over token creation by implementing ITokenSigningService - that's where you should start your research.

**Comment by kalanbates** [comment]
>How do you expect that claim to de-serialize "on the other end" - are you using .NET ?

I'm deserializing with a javascript (angular) client.  

>You can take control over token creation by implementing ITokenSigningService
I'll dig in and see what I can produce.  This was going to be my item for today but my priorities were swapped on me first thing, so I'll report back to this item sometime tomorrow.  

Thanks for the information.

**Comment by leastprivilege** [comment]
progress?

**Comment by kalanbates** [comment]
Not yet. Production Support pulled me away for a bit longer than expected, but digging in as soon as I finish this comment. (Came here to check if anyone else might have added information since the last time I updated)

**Comment by leastprivilege** [comment]
I just added a branch with enhanced token creation code

https://github.com/IdentityServer/IdentityServer3/tree/EnhancedTokenSigningService

create your claim with a ClaimValuesType of JsonArray and give it a try - like this:

 https://github.com/IdentityServer/IdentityServer3/blob/EnhancedTokenSigningService/source/Host.Configuration/Config/Users.cs#L40

**Comment by kalanbates** [comment]
Gah!!! You beat me to it. lol  Good to know that we were independently using similar approaches (although mine wasn't quite finished.)

 a few of the differences between our approaches:
1) I was using "json" as the claim type and then letting my JsonConverter interpret whether it needed to be cast as a JArray or a JObject.
2) My change was a bit more invasive because I didn't think to assign to JsonExtensions.Serializer and "just let it fly." I was creating my own JwtSecurityTokenHandler wrapper, wrapping JwtSecurityToken with customizations then overriding EncodedHeader and EncodedPayload so that I could dig into SerializeToJson and call `JsonConvert.SerializeToObject` rather than the framework default's `JsonExtensions.SerializeToJson`

I was still getting escaped JSON kicked out of the Serializer, so there was something left to tinker with.
I plugged your new TokenSigningService into my implementation and it kicked out exactly what I needed.

```
 "synchcode": "69a75f26-623e-4e6f-ab88-5e32b06e3bbd",
  "correlator": "dbb2e477-c2bf-44f4-82f1-821abc7766f6",
  "roles": [
    {
      "Code": "FOO",
      "Value": "ROLE_LEGACY_FOO"
    }]
```

Thanks for this!

**Comment by kalanbates** [comment]
never mind.  PEBCAK
